parl/algorithm_zoo/simple_algorithms.py

```diff
@@ -0,0 +1,161 @@
+        """
+        super(SimpleAC, self).__init__(model_func, gpu_id)
+        self.num_actions = num_actions
+        self.discount_factor = discount_factor
+        self.min_exploration = min_exploration
+        ## create some layers for later use
+        self.mlp = Feedforward([layers.fc(**c) for c in mlp_layer_confs])
+        self.policy_layer = layers.fc(num_actions, act='softmax')
+        self.value_layer = layers.fc(1)
+
+    def _predict(self, policy_states):
+        assert len(policy_states) == 1
+        policy = self.policy_layer(self.mlp(policy_states.values()[0]))
+        policy = comf.sum_to_one_norm_layer(policy + self.min_exploration)
+        return dict(action=comf.discrete_random(policy))
+
+    def _learn(self, policy_states, next_policy_states, actions, rewards):
+        assert len(actions) == 1
+        assert len(policy_states) == 1
+        assert len(next_policy_states) == 1
+        assert len(rewards) == 1
+        action = actions.values()[0]
+        policy_state = policy_states.values()[0]
+        next_policy_state = next_policy_states.values()[0]
+        reward = rewards.values()[0]
+
+        referent_value = self.ref_alg.__value(next_policy_state)
+        referent_value.stop_gradient = True
+
+        critic_value = reward + self.discount_factor * referent_value
+        td_error = critic_value - self.__value(policy_state)
+        value_cost = layers.square(td_error)
+
+        policy = self.policy_layer(self.mlp(policy_state))
+        policy = comf.sum_to_one_norm_layer(policy + self.min_exploration)
+        pg_cost = layers.cross_entropy(input=policy, label=action)
+        return dict(cost=value_cost + pg_cost * td_error)
+
+    def __value(self, policy_state):
+        return self.value_layer(self.mlp(policy_state))
+
+
+class SimpleQ(Algorithm):
+    """
+    A simple Q-learning that has a feedforward policy network and a single discrete action.
+    """
+
+    def __init__(self,
+                 model_func,
+                 num_actions,
+                 mlp_layer_confs,
+                 gpu_id=-1,
+                 discount_factor=0.99,
+                 min_exploration=0.01,
+                 update_ref_interval=100):
+        """
+        num_actions: the number of discrete actions
+        mlp_layer_confs: a list of fc configs
+        """
+        super(SimpleQ, self).__init__(model_func, gpu_id)
+        self.num_actions = num_actions
+        self.discount_factor = discount_factor
+        self.min_exploration = min_exploration
+        self.mlp_layer_confs = mlp_layer_confs
+        self.gpu_id = gpu_id
+        assert update_ref_interval > 0
+        self.update_ref_interval = update_ref_interval
+        self.total_batches = 0
+        ## create some layers for later use
+        self.mlp = Feedforward([layers.fc(**c) for c in mlp_layer_confs])
+
+    def init(self):
+        """
+        We create a new copy
+        """
+        self.ref_alg = SimpleQ(
```

Comment:
how about adding clone() function for Model to clone model? That might be easier to understand.

Reply:
Are you talking about clone() function for Algorithm or Model? I made a change such that there is no more "model_func" for the Algorithm __init__(). However, the copy between two algorithms is still tedious.